src/main/java/org/jboss/threads/EnhancedQueueExecutor.java

```diff
@@ -1706,17 +1714,30 @@ private int tryExecute(final Runnable runnable) {
         TaskNode node = null;
         for (;;) {
             tailNext = tail.getNext();
+            // tail is already consumed
+            if (tailNext == tail) {
+                if (UPDATE_STATISTICS) spinMisses.increment();
+                JDKSpecific.onSpinWait();
+                // retry with new tail(snapshot)
+                tail = this.tail;
+                continue;
+            }
+            assert tailNext != tail;
             if (tailNext instanceof TaskNode) {
-                TaskNode tailNextTaskNode;
-                do {
-                    if (UPDATE_STATISTICS) spinMisses.increment();
-                    tailNextTaskNode = (TaskNode) tailNext;
-                    // retry
-                    tail = tailNextTaskNode;
-                    tailNext = tail.getNext();
-                } while (tailNext instanceof TaskNode);
-                // opportunistically update for the possible benefit of other threads
-                if (UPDATE_TAIL) compareAndSetTail(tail, tailNextTaskNode);
+                if (UPDATE_TAIL) {
```

Comment:
@carterkozak I've used the same approach of you on https://github.com/jbossas/jboss-threads/pull/73#issuecomment-616953067, but saving to walk the list if UPDATE_TAIL won't validate (using the compareAndSet) that we're lagging behind `this.tail`.
I've changed this code path as well because the walking algorithm would have changed due to the chance to get  a recursive next pointer when a node is consumed.

Reply:
Unfortunately I think we have to walk the list to avoid blocking on execute. I don't think UPDATE_TAIL should ever be disabled, see https://github.com/jbossas/jboss-threads/pull/72#discussion_r412149895

Reply:
I'm answering directly there :)

Reply:
@dmlloyd @carterkozak Regardless the way I've written this specific code path: the GC and the assertion issue on https://github.com/jbossas/jboss-threads/pull/74#issuecomment-617026360 are still valid; Probably will rebase this code path on what will be merged after  https://github.com/jbossas/jboss-threads/pull/73